stan-dev/stan
# Stan refactor

#### Submisison Checklist
#### Summary

This pull request accomplishes two major things:
1. In the `src/stan/services` folder (the `stan::services` namespace), we now have functions that should be called by the interfaces.
2. Removes a lot of the old code.

It does this while maintaining all current behavior.
#### Intended Effect

The intended effect is to make the downstream interfaces easier to implement and maintain.

The code is much cleaner. Each of the functions in `src/stan/services` excluding the `src/stan/services/util` folder are meant to be called by RStan, PyStan, and CmdStan. This will also make it much easier to deal with the upcoming template build issue we will have with the introduction of forward-mode and RHMC.
#### How to Verify

(I'm sorry -- the diff is large.)
1. All tests should pass. (The upstream tests will fail due to CmdStan `develop` branch not having changes necessary, but I'll submit a pull request on the CmdStan branch.)
2. Take a look at all the new functions in `src/stan/services`. I've covered all the use cases of CmdStan, PyStan, and RStan. See side effects below.
3. There were a couple of additional utilities outside of `src/stan/services` that were added. Take a look.

If there are some oddly implemented things in there, most weird implementations are due to keeping the existing behavior intact. This way, we can roll these changes out with a minor release and keep moving forward.
#### Side Effects

CmdStan, RStan, and PyStan will break under the current branch, but will maintain the same behavior.

I have a branch of CmdStan that works. Allen has a branch of PyStan that's almost working. I have a branch of RStan that's almost working.
#### Documentation

The C++ code may need some more doc.

If anyone wants to help, there's that and adding more tests.
#### Reviewer Suggestions

This one's big, so it'd be helpful to have a few eyes on it.

@ariddell, @bgoodri: since this will effect both of you directly.
@sakrejda: since you've been looking at this for a while.
@bob-carpenter, @betanalpha, @rtrangucci: if you've got the time.

Although I'm confident that this behaves identically, this one shouldn't be taken lightly.
#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
